packages/components/src/autocomplete/index.js

```diff
@@ -183,40 +184,46 @@ function useAutocomplete( {
 		setAutocompleterUI( null );
 	}
 
-	function announce( options ) {
-		if ( ! debouncedSpeak ) {
-			return;
-		}
-		if ( !! options.length ) {
-			debouncedSpeak(
-				sprintf(
-					/* translators: %d: number of results. */
-					_n(
-						'%d result found, use up and down arrow keys to navigate.',
-						'%d results found, use up and down arrow keys to navigate.',
+	const announce = useCallback(
+		( options ) => {
+			if ( ! debouncedSpeak ) {
+				return;
+			}
+			if ( !! options.length ) {
+				debouncedSpeak(
+					sprintf(
+						/* translators: %d: number of results. */
+						_n(
+							'%d result found, use up and down arrow keys to navigate.',
+							'%d results found, use up and down arrow keys to navigate.',
+							options.length
+						),
 						options.length
 					),
-					options.length
-				),
-				'assertive'
-			);
-		} else {
-			debouncedSpeak( __( 'No results.' ), 'assertive' );
-		}
-	}
+					'assertive'
+				);
+			} else {
+				debouncedSpeak( __( 'No results.' ), 'assertive' );
+			}
+		},
+		[ debouncedSpeak ]
+	);
 
 	/**
 	 * Load options for an autocompleter.
 	 *
 	 * @param {Array} options
 	 */
-	function onChangeOptions( options ) {
-		setSelectedIndex(
-			options.length === filteredOptions.length ? selectedIndex : 0
-		);
-		setFilteredOptions( options );
-		announce( options );
-	}
+	const onChangeOptions = useCallback(
+		( options ) => {
+			setSelectedIndex(
+				options.length === filteredOptions.length ? selectedIndex : 0
+			);
+			setFilteredOptions( options );
+			announce( options );
+		},
+		[ announce, filteredOptions.length, selectedIndex ]
+	);
```

Comment:
So... I'm starting to think about when it's better to use a `useCallback` vs. putting the callback in a `ref`.

In this case, the `ref` approach would be to change this

https://github.com/WordPress/gutenberg/blob/982cfc775d15b840c08a06e25cc72c69587185d7/packages/components/src/autocomplete/autocompleter-ui.js#L40-L42

to something like this:

```js
		const onChangeOptionsRef = useRef( onChangeOptions );
		onChangeOptionsRef.current = onChangeOptions;

		useLayoutEffect( () => {
			onChangeOptionsRef.current( items );
		}, [ items ] );
```

The reason I think this approach is preferable here is because, it should be `AutocompleterUI`'s responsibility that its own effects work correctly. If we instead address it with a `useCallback` in the consumer, then all potential consumers will need to do the same thing, which is hard to know about.

I might be missing something though. What do you think?